### DISCUSSION TITLE
originalEntity in afterFlush hook destructures embeddable fields

### DISCUSSION THREAD

**Discussion by njeirath** [mikro-orm/mikro-orm#5756] [General]
### Describe the bug

We're attempting to use the afterFlush hook to compare the original entity with the updated entity and perform some actions if certain fields in the entity have been updated. This typically works fine but our entity has an embeddable in it and it appears that in the original entity in the change set has its properties destructured while the entity does not.

### Reproduction

```ts
import { Embeddable, Embedded, Entity, EventSubscriber, FlushEventArgs, MikroORM, PrimaryKey, Property } from '@mikro-orm/sqlite';

@Embeddable()
class Img {
  @Property()
  url: string;

  @Property()
  thumbnail: string;

  constructor(url: string, thumbnail: string) {
    this.url = url;
    this.thumbnail = thumbnail;
  }
}

@Entity()
class User {

  @PrimaryKey()
  id!: number;

  @Property()
  name: string;

  @Embedded(() => Img, {nullable: true})
  image?: Img;

  constructor(name: string) {
    this.name = name;
  }
}

export class PersonSubscriber implements EventSubscriber<User> {
  constructor(private messageSender: any) {}

  async afterFlush(args: FlushEventArgs): Promise<void> {
    const messages = args.uow
      .getChangeSets()
      .forEach((cs) => this.messageSender(cs.originalEntity, cs.entity))
  }
}

let orm: MikroORM;
const mockSender = jest.fn();

beforeAll(async () => {
  orm = await MikroORM.init({
    dbName: ':memory:',
    entities: [User, Img],
    subscribers: [new PersonSubscriber(mockSender)],
    debug: ['query', 'query-params'],
    allowGlobalContext: true, // only for testing
  });
  await orm.schema.refreshDatabase();
});

afterAll(async () => {
  await orm.close(true);
});

test('basic CRUD example', async () => {
  const user = new User('Name');
  user.image = new Img('url', 'thumb')
  await orm.em.persistAndFlush(user);

  expect(mockSender).toHaveBeenCalledTimes(1);
  expect(mockSender).toHaveBeenCalledWith(undefined, 
    {id: 1, name: 'Name', image: {url: 'url', thumbnail: 'thumb'}});

  user.image.thumbnail = 'thumb2'
  await orm.em.flush()

  expect(mockSender).toHaveBeenCalledTimes(2);
  expect(mockSender).toHaveBeenCalledWith(
    {id: 1, name: 'Name', image: {url: 'url', thumbnail: 'thumb'}}, 
    {id: 1, name: 'Name', image: {url: 'url', thumbnail: 'thumb2'}}); //FAILS HERE
  
});
```

The above reproduction fails because image in the originalEntity does not exist and instead we get `image_url` and `image_thumbnail` properties separately:
```
              Object {
                "id": 1,
            -   "image": Object {
            -     "thumbnail": "thumb",
            -     "url": "url",
            -   },
            +   "image_thumbnail": "thumb",
            +   "image_url": "url",
                "name": "Name",
              },
              {"id": 1, "image": {"thumbnail": "thumb2", "url": "url"}, "name": "Name"},
```

### What driver are you using?

@mikro-orm/postgresql

### MikroORM version

5.7.14 (also fails on latest version)

### Node.js version

v20.15.0

### Operating system

MacOS

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Slack](https://join.slack.com/t/mikroorm/shared_invite/enQtNTM1ODYzMzM4MDk3LWM4ZDExMjU5ZDhmNjA2MmM3MWMwZmExNjhhNDdiYTMwNWM0MGY5ZTE3ZjkyZTMzOWExNDgyYmMzNDE1NDI5NjA).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
This is basically the same as #5621 (even harder actually, since you would like to see the internals change, not just the result of `qb.execute()`) and I consider that a wontfix, because it would mean fundamental changes to the diffing mechanism (on multiple levels), and I honestly do not see an easy way to support this.

**Comment by njeirath** [comment]
@B4nan thanks for the reply and I understand it'd be a major change to support this. I saw your comments in #5621 refers to using virtual entities but it looks like that only applies to mapping a query to a virtual entity. I'm curious if there's a function somewhere in the ORM internals that I might be able to use to perform the mapping from, what I'm assuming, are raw originalEntity DB fields to my entity including the embeddable or if it's something I'll just have to manually do?

**Reply by B4nan** [reply]
I don't think we have such a thing, the mapping part for embeddables is done only when hydrating the entities, and there is no "old entity state", the only entity state is the one that is being mutated. 

I actually did quite a lot of work (spent couple of weeks on it) for #5621 and I would still like to revisit that at some point, but it's such a deep refactoring that this won't likely happen before v7.

**Reply by njeirath** [reply]
No worries, I'll handle it manually. As always, thanks for the help @B4nan
